### PR TITLE
FMComms234 Source/Sink fixes

### LINKED ISSUE #5476: FMComms 2/3/4 Sources and Sinks not working
### What happened?

With FMComms 2/3/4 Sources or Sinks, the grc file would return an error during generation.
I compared FMComms to Pluto SDR, it seems the FMComms components lack len_tag_key configuration fields in properties.

Error message below.

### System Information

OS: Ubuntu 20.04 LTS
GR Installation Method: Source

### GNU Radio Version

3.10-git (master)

### Specific Version

3.10.0.0

### Steps to Reproduce the Problem

Add an FMComms 2/3/4 source or sink and click "Generate the flow graph" in the gnuradio companion.

### Relevant log output

```
Generate Error: (NameError("'len_tag_key' is not defined"), 'iio.fmcomms2_sink_${type.type}(${uri}, [tx1_en, tx2_en], ${buffer_size}, ${cyclic})\nself.${id}.set_len_tag_key(${len_tag_key})\nself.${id}.set_bandwidth(${bandwidth})\nself.${id}.set_frequency(${frequency})\nself.${id}.set_samplerate(${samplerate})\nself.${id}.set_attenuation(0, ${attenuation1})\nself.${id}.set_filter_params(${filter_source}, ${filter}, ${fpass}, ${fstop})\n')
```

### PR DESCRIPTION
**Pull request by tfcollins**
# Pull Request Details

PR fixes #5476 for both the FMComms234 source and sink blocks.

## Description

The fixes update the GRC files to use the correct APIs for 1RX/2RX 1TX/2TX configurations that interface with AD9361/AD9363/AD9364 based systems using the IIO drivers. The core iio_device_{source/sink} blocks have been updated to allow inherited blocks to better define tags to be generated/processed when using the packet_len APIs. Unrelated to #5476 was the incorrect behavior of the incorrect number of channels being processed for these tags. All of these have been addressed and tested.

## Related Issue

#5476 

## Which blocks/areas does this affect?

- fmcomms2_source (PlutoSDR, FMComms 2/3/4)
- fmcomms2_sink (PlutoSDR, FMComms 2/3/4)
- iio_device_source
- iio_device_sink

## Testing Done

Flowgraphs have been generated and run without error for device combinations with source and sink blocks:
- ADRV9361-Z7035 (Equivalent to FMComms2 and FMComms3)
- PlutoSDR (Equivalent to FMComms4)

## Checklist

- [X] I have read the [CONTRIBUTING document](https://github.com/gnuradio/gnuradio/blob/main/CONTRIBUTING.md).
- [X] I have squashed my commits to have one significant change per commit. 
- [X] I [have signed my commits before making this PR](https://github.com/gnuradio/gnuradio/blob/main/CONTRIBUTING.md#dco-signed)
- [X] My code follows the code style of this project. See [GREP1.md](https://github.com/gnuradio/greps/blob/main/grep-0001-coding-guidelines.md).
- [ ] I have updated [the documentation](https://wiki.gnuradio.org/index.php/Main_Page#Documentation) where necessary.
- [ ] I have added tests to cover my changes, and all previous tests pass.

### CODE REVIEW

File: gr-iio/lib/fmcomms2_sink_impl.cc [hunk 1]
```diff
@@ -338,7 +341,7 @@ int fmcomms2_sink_impl<gr_complex>::work(int noutput_items,
 
     for (size_t i = 0; i < input_items.size(); i++) {
         auto in = static_cast<const gr_complex*>(input_items[i]);
-        if (noutput_items > (int)d_device_bufs[i].size()) {
+        if (noutput_items > (int)d_device_bufs[2 * i].size()) {
```
Comment (mormj) [inline gr-iio/lib/fmcomms2_sink_impl.cc:344]: Looks like this bug (sorry for this one) exists in the int16_t work specialization above as well

**Review by Yamakaja — COMMENTED** [review]
Did you run `clang-format`?

File: gr-iio/lib/fmcomms2_source_impl.cc [hunk 2]
```diff
@@ -334,7 +334,7 @@ template <typename T>
 void fmcomms2_source_impl<T>::set_gain_mode(size_t chan, const std::string& mode)
 {
     bool is_fmcomms4 = !iio_device_find_channel(phy, "voltage1", false);
-    if ((!is_fmcomms4 && chan > 0) || chan > 1) {
+    if ((is_fmcomms4 && chan > 0) || chan > 1) {
```
Comment (Yamakaja) [inline gr-iio/lib/fmcomms2_source_impl.cc:340]: Shouldn't we be checking for the `voltage2` channel to determine if this is a single-channel device? (`[voltage0, voltage1] ^= [ch0.real,ch0.imag]` iirc)

File: gr-iio/lib/fmcomms2_sink_impl.cc [hunk 3]
```diff
@@ -126,6 +126,9 @@ fmcomms2_sink_impl<T>::fmcomms2_sink_impl(iio_context* ctx,
     }
     d_float_r.resize(s_initial_device_buf_size);
     d_float_i.resize(s_initial_device_buf_size);
+
+    // Tell tagger in device_sink_impl::work that we are using a less inputs
```
[the remark this answers is not in the document]
Comment (tfcollins) [reply gr-iio/lib/fmcomms2_sink_impl.cc:130]: Which part? Its just referencing this https://github.com/tfcollins/gnuradio/blob/tfcollins/fmcomms-fix/gr-iio/lib/device_source_impl.cc#L295

File: gr-iio/lib/fmcomms2_source_impl.cc (hunk 2 above)
Comment (tfcollins) [reply gr-iio/lib/fmcomms2_source_impl.cc:340]: If we were checking the DMAs but this is w.r.t. to the PHY.

ADRV9361-Z7035 (FMComms2/3 are the same):
```
❯ iio_attr -u ip:analog.local -c ad9361-phy
dev 'ad9361-phy', channel 'altvoltage1', id 'TX_LO' (output), found 8 channel-specific attributes
dev 'ad9361-phy', channel 'voltage1' (output), found 10 channel-specific attributes
dev 'ad9361-phy', channel 'voltage0' (input), found 15 channel-specific attributes
dev 'ad9361-phy', channel 'voltage1' (input), found 15 channel-specific attributes
dev 'ad9361-phy', channel 'voltage3' (output), found 8 channel-specific attributes
dev 'ad9361-phy', channel 'altvoltage0', id 'RX_LO' (output), found 8 channel-specific attributes
dev 'ad9361-phy', channel 'voltage2' (output), found 8 channel-specific attributes
dev 'ad9361-phy', channel 'temp0' (input), found 1 channel-specific attributes
dev 'ad9361-phy', channel 'voltage0' (output), found 10 channel-specific attributes
dev 'ad9361-phy', channel 'voltage2' (input), found 13 channel-specific attributes
dev 'ad9361-phy', channel 'out' (input), found 1 channel-specific attributes
```

Pluto (FMComms4 is the same):
```
❯ iio_attr -u ip:pluto.local -c ad9361-phy
dev 'ad9361-phy', channel 'altvoltage1', id 'TX_LO' (output), found 8 channel-specific attributes
dev 'ad9361-phy', channel 'voltage0' (input), found 15 channel-specific attributes
dev 'ad9361-phy', channel 'voltage3' (output), found 8 channel-specific attributes
dev 'ad9361-phy', channel 'altvoltage0', id 'RX_LO' (output), found 8 channel-specific attributes
dev 'ad9361-phy', channel 'voltage2' (output), found 8 channel-specific attributes
dev 'ad9361-phy', channel 'temp0' (input), found 1 channel-specific attributes
dev 'ad9361-phy', channel 'voltage0' (output), found 10 channel-specific attributes
dev 'ad9361-phy', channel 'voltage2' (input), found 13 channel-specific attributes
dev 'ad9361-phy', channel 'out' (input), found 1 channel-specific attributes
```

File: gr-iio/lib/fmcomms2_sink_impl.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (Yamakaja) [reply gr-iio/lib/fmcomms2_sink_impl.cc:130]: "we are using **a** less inputs". (Same goes for the source_impl.cc)

**Review by mormj — APPROVED** [review]
Thanks so much @tfcollins !!

**Pull request merged**